planner/core/logical_plan_builder.go

```diff
@@ -585,6 +602,7 @@ func (b *PlanBuilder) buildProjection(p LogicalPlan, fields []*ast.SelectField,
 	oldLen := 0
 	for _, field := range fields {
 		newExpr, np, err := b.rewrite(field.Expr, p, mapper, true)
+		newExpr = eliminateIfNullOnNonNullColumn(p, newExpr)
```

Comment:
We can add a TODO here: Do this optimization automatically when creating new scalar function.